src/stream/src/executor/temporal_join.rs

```diff
@@ -0,0 +1,214 @@
+use crate::cache::ManagedLruCache;
+use crate::common::StreamChunkBuilder;
+
+pub struct TemporalJoinExecutor<S: StateStore> {
+    left: Box<dyn Executor>,
+    right: TemporalSide<S>,
+    right_stream: Box<dyn Executor>,
+    left_join_keys: Vec<usize>,
+    right_join_keys: Vec<usize>,
+    output_indices: Vec<usize>,
+    schema: Schema,
+    chunk_size: usize,
+}
+
+struct TemporalSide<S: StateStore> {
+    source: StorageTable<S>,
+    cache: ManagedLruCache<OwnedRow, Option<OwnedRow>>,
+}
+
+impl<S: StateStore> TemporalSide<S> {
+    async fn lookup(
+        &mut self,
+        key: impl Row,
+        epoch: EpochPair,
+    ) -> StreamExecutorResult<Option<OwnedRow>> {
+        let key = key.into_owned_row();
+        Ok(match self.cache.get(&key) {
+            Some(res) => res.clone(),
+            None => {
+                let res = self
+                    .source
+                    .get_row(key.clone(), HummockReadEpoch::NoWait(epoch.curr))
+                    .await?;
+                self.cache.put(key, res.clone());
+                res
+            }
+        })
+    }
+
+    fn update(&mut self, payload: Vec<StreamChunk>, join_keys: &[usize]) {
+        payload.iter().flat_map(|c| c.rows()).for_each(|(op, row)| {
+            let key = row.project(join_keys).into_owned_row();
+            if let Some(value) = self.cache.get_mut(&key) {
+                match op {
+                    Op::Insert | Op::UpdateInsert => *value = Some(row.into_owned_row()),
+                    Op::Delete | Op::UpdateDelete => *value = None,
+                };
+            }
+        })
+    }
+}
+
+enum InternalMessage {
+    Barrier(Vec<StreamChunk>, Barrier),
+    Chunk(StreamChunk),
+}
+
+#[try_stream(ok = StreamChunk, error = StreamExecutorError)]
+pub async fn chunks_until_barrier(stream: impl MessageStream, expected_barrier: Barrier) {
+    #[for_await]
+    for item in stream {
+        match item? {
+            Message::Watermark(_) => {
+                todo!("https://github.com/risingwavelabs/risingwave/issues/6042")
+            }
+            Message::Chunk(c) => yield c,
+            Message::Barrier(b) if b.epoch != expected_barrier.epoch => {
+                return Err(StreamExecutorError::align_barrier(expected_barrier, b));
+            }
+            Message::Barrier(_) => return Ok(()),
+        }
+    }
+}
+
+// Align the left and right inputs according to their barriers,
+// such that in the produced stream, an aligned interval starts with
+// `InternalMessage::Barrier(right_chunks, barrier)` and followed by
+// any number of `InternalMessage::Chunk(left_chunk)`.
+// Notice that we collect all chunks from the right side up front,
+// prioritizing before any chunks from the left side in the same interval.
+#[try_stream(ok = InternalMessage, error = StreamExecutorError)]
+async fn align_input(left: Box<dyn Executor>, right: Box<dyn Executor>) {
```

Comment:
I'm not sure if using the hash join implementation would work, as I understand it does not guarantee any ordering of left and right chunks within two barriers.
Suppose the left chunk wish to join with `a = 1`, and the right chunk insert a row with `a = 1`, then the result would then dependent on whether we process the left or right chunk first.
I was trying to achieve something more like the alignment in lookup join where the arrangement chunks always come before streaming chunks, hence making sure the join result is deterministic.

I agree that the current implementation seems to make a barrier occur before the data it guards, but maybe this can be fixed by changing the convention to have `InternalMessage::Barrier` always occur after `InternalMessage::Chunk`, making the left side always joining with the committed version of the right side?
If we wish to achieve a more real-time join, then using the epoch from a prior barrier may also work?

Reply:
> I agree that the current implementation seems to make a barrier occur before the data it guards, but maybe this can be fixed by changing the convention to have InternalMessage::Barrier always occur after InternalMessage::Chunk, making the left side always joining with the committed version of the right side

+1 for this. Let's implement the first version of temporal join to lookup the committed version of the right side.